README.md
# tcp-server

[![Build Status](https://travis-ci.org/nshimaza/tcp-server.svg?branch=master)](https://travis-ci.org/nshimaza/tcp-server)

A Simple Plain TCP Server.

Boot a simple, plain, general purpose TCP server with automatic hierarchical thread management.
It accepts a user provided handler and forks dedicated thread which listens given TCP port number.
Every time the listening thread accept a new connection from peer, it forks new thread dedicated for
each connection then calls user provided handler.

When server shutdown, all threads including listening thread, handler threads and their children
automatically killed.  You don't have to close the socket by your own.  This module automatically
close the socket on handler exit regardless if it exited normally or it was killed.
You can install your own cleanup routine using finally or bracket.


### Example

```haskell
import           Control.Monad        (forever, unless)
import qualified Data.ByteString      as B (null)
import qualified Data.ByteString.Lazy as BL (fromStrict)

import           Network.TcpServer

newEchoServer :: IO TcpServer
newEchoServer = newTlsServer 8443 echoServerHandler

echoServerHandler :: TransportHandler IO
echoServerHandler _ peer = go
  where
    go = do
        msg <- transportRecv peer
        unless (B.null msg) $ do
            transportSend peer $ BL.fromStrict msg
            go
```


ChangeLog.md
# Changelog for tcp-server

## Unreleased changes


test/Network/TcpServerSpec.hs
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Network.TcpServerSpec where

import           Prelude                    hiding (null)

import           Control.Concurrent         (threadDelay)
import           Control.Exception          (IOException, bracket, catch)
import           Control.Monad              (unless, void, zipWithM_)
import           Data.ByteString            (null, singleton)
import qualified Data.ByteString.Char8      as BC8 (pack)
import           Data.ByteString.Lazy       (fromStrict)
import qualified Data.ByteString.Lazy.Char8 as BLC8 (pack)
import           Data.Default.Class
import           Data.Foldable              (for_, traverse_)
import           Data.Maybe                 (isJust)
import           Data.Monoid                ((<>))
import           Data.Traversable           (traverse)
import           Network.Socket             (Family (..), ShutdownCmd (..),
                                             SockAddr (..), Socket,
                                             SocketType (..), close, connect,
                                             defaultProtocol, isConnected,
                                             isWritable, shutdown, socket,
                                             tupleToHostAddress)
import           Network.Socket.ByteString  (recv, sendAll)
import           Network.TLS                (ClientParams (..), Context,
                                             Credentials (..), Logging (..),
                                             ServerParams (..), Shared (..),
                                             ValidationCache (..),
                                             ValidationCacheResult (..), bye,
                                             contextClose,
                                             contextGetInformation,
                                             contextHookSetLogging, contextNew,
                                             credentialLoadX509FromMemory,
                                             handshake, recvData, sendData,
                                             sharedCredentials,
                                             supportedCiphers,
                                             supportedClientInitiatedRenegotiation)
import           Network.TLS.Extra.Cipher   (ciphersuite_default)

import           Test.Hspec

import           Network.TcpServer

listenPort = 8080

helloWorldMessage = "hello, world"

startTcpServer :: TransportHandler -> IO TcpServer
startTcpServer handler = do
    svr <- newTcpServer listenPort handler
    waitListen svr
    return svr

connToTcpServer :: IO Socket
connToTcpServer = do
            sk <- socket AF_INET Stream defaultProtocol
            connect sk . SockAddrInet listenPort $ tupleToHostAddress (127,0,0,1)
            return sk

closeTcp :: Socket -> IO ()
closeTcp sk = shutdown sk ShutdownSend `catch` (\(_ :: IOException) -> return ()) >> close sk

loggingHooks = def { loggingPacketSent = \packet -> putStrLn ("C: PacketSent " ++ show packet)
                   , loggingPacketRecv = \packet -> putStrLn ("C: PacketRecv " ++ show packet)
                   , loggingIOSent = \io -> putStrLn ("C: IOSent " ++ show io)
                   , loggingIORecv = \header io -> putStrLn ("C: IORecv header:" ++ show header ++ " io:" ++ show io)
                   }

certificatePem = "-----BEGIN CERTIFICATE-----\n\
                 \MIIEcTCCA1mgAwIBAgIJAIxzVlmJ0tBTMA0GCSqGSIb3DQEBBQUAMIGBMQswCQYD\n\
                 \VQQGEwJKUDEOMAwGA1UECBMFVG9reW8xDjAMBgNVBAcTBVRva3lvMQ4wDAYDVQQK\n\
                 \EwVEdW1teTEOMAwGA1UECxMFRHVtbXkxEDAOBgNVBAMTB2ZvcnRlc3QxIDAeBgkq\n\
                 \hkiG9w0BCQEWEWR1bW15QGV4YW1wbGUuY29tMB4XDTE3MDYyMzEwMzMxNVoXDTE4\n\
                 \MDYyMzEwMzMxNVowgYExCzAJBgNVBAYTAkpQMQ4wDAYDVQQIEwVUb2t5bzEOMAwG\n\
                 \A1UEBxMFVG9reW8xDjAMBgNVBAoTBUR1bW15MQ4wDAYDVQQLEwVEdW1teTEQMA4G\n\
                 \A1UEAxMHZm9ydGVzdDEgMB4GCSqGSIb3DQEJARYRZHVtbXlAZXhhbXBsZS5jb20w\n\
                 \ggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDfX0fAvpJkt9kNN9OmYZU8\n\
                 \M2O9iogVTbSj0nTCd3aID9aJcdfnZECa0fROO/C3Cjbnm80nUHeskuNmA8Q3oc9C\n\
                 \JOXe+21yjR6A9kP7gLqXkTHmWcZhzgiZxLxdUZTL6m7Gc6UTSauwJrfHsKtfYLcM\n\
                 \ux4LmFq2KtXECCwahTz6JkBqBFkcxpf+Gn3ccJPYcpb/OzuytgwJiuQOOduQ2E+2\n\
                 \2HsHxLW7JneovMjCAVJCCrs9+cjp+O2l4U97JIOqFWLjVt+qB/kjEzFnspPgCl+s\n\
                 \fltls4VZX6BQQofk+UihfxEIjGNHTQcBuD098tSqdjk9vlmutre7B2C1aBAQk3MV\n\
                 \AgMBAAGjgekwgeYwHQYDVR0OBBYEFHSA1t0VgSrJC/fs8h6qWtcHX9NkMIG2BgNV\n\
                 \HSMEga4wgauAFHSA1t0VgSrJC/fs8h6qWtcHX9NkoYGHpIGEMIGBMQswCQYDVQQG\n\
                 \EwJKUDEOMAwGA1UECBMFVG9reW8xDjAMBgNVBAcTBVRva3lvMQ4wDAYDVQQKEwVE\n\
                 \dW1teTEOMAwGA1UECxMFRHVtbXkxEDAOBgNVBAMTB2ZvcnRlc3QxIDAeBgkqhkiG\n\
                 \9w0BCQEWEWR1bW15QGV4YW1wbGUuY29tggkAjHNWWYnS0FMwDAYDVR0TBAUwAwEB\n\
                 \/zANBgkqhkiG9w0BAQUFAAOCAQEAqQoEYsAJ1wIECV4jdA/QPPURlPzthGV8I//I\n\
                 \xtkBojoH9r/q0VCTaKhxmEwS3wU8TDuH1SM3CJp1AYCl812OWuo4khBeMTi1ltxl\n\
                 \71yWkGqYQxrqUNvfr4Oy12irv09D4nsSmRPbWHz/RPLFKIiCAyLgBIq5HhO8KzyP\n\
                 \lNyzsUIPs55xdCTLKH+O48gg80RXNRLOk3XTZuNIHKtbZgVE9k4Vo+vW8/SfDdmK\n\
                 \8CqUgXyU05rbMUW5bJjBELoARxW1+K730wo8POhcur2r5XC574RLfCdInWQqFLgd\n\
                 \bnqHotrGDXbBuKQcobzRQTPWZCAuemsI2S5ZxIbhdwu3TH5FJA==\n\
                 \-----END CERTIFICATE-----"

keyPem = "-----BEGIN RSA PRIVATE KEY-----\n\
         \MIIEpQIBAAKCAQEA319HwL6SZLfZDTfTpmGVPDNjvYqIFU20o9J0wnd2iA/WiXHX\n\
         \52RAmtH0Tjvwtwo255vNJ1B3rJLjZgPEN6HPQiTl3vttco0egPZD+4C6l5Ex5lnG\n\
         \Yc4ImcS8XVGUy+puxnOlE0mrsCa3x7CrX2C3DLseC5hatirVxAgsGoU8+iZAagRZ\n\
         \HMaX/hp93HCT2HKW/zs7srYMCYrkDjnbkNhPtth7B8S1uyZ3qLzIwgFSQgq7PfnI\n\
         \6fjtpeFPeySDqhVi41bfqgf5IxMxZ7KT4ApfrH5bZbOFWV+gUEKH5PlIoX8RCIxj\n\
         \R00HAbg9PfLUqnY5Pb5Zrra3uwdgtWgQEJNzFQIDAQABAoIBAGN4CnvKYWZ4eC9q\n\
         \rNVqnwov2DqPdgoGkGpjTZtLl3z/kImkKIDiI+LNm/06s/bOlaSNvM3FB4XR43GK\n\
         \O9aKN8E8rSCe1h4suWi0QZG9wVm9TntFfxIOrcXL1hW4PoqP2sSEuv3b0JigpTZH\n\
         \dgkDCICi0r5XGg4FBSxGd11X989sZCe978mz3sCxZIkADk6IpKqLGOPd3iLuwZS+\n\
         \Pxa72pFv4QqzyQ8T00pgFflc1Szmj8CCBrgO7hnJo37SI+ggOM/GAXoKpwBkqUgD\n\
         \Rlw0sVC8zbXYNu278uGCdghYb14vFvH7P+ZLUXAhq+g9T6BqYp1ktwUCyDjBqJKe\n\
         \UQ9vWAECgYEA96VAHlU9xTW6OdsUZ+F6neAMqVExUFtj5G1mLsVr+m2YUH1UjDat\n\
         \Dc15TQPZ+zHA2rhlu0pOfviYJ3HSD2qwg8zsCt7yIS3rPOqsPRn3TlTNsF1hLSCC\n\
         \rdhleDB0l2tRb2eDHfkzStm/tHJi74wQV4feJJtDGEgKtzVijcDvrKMCgYEA5uhl\n\
         \tRkvxElbcqanMgMeQscvGmPYXVuDT1Eb7AyXIPWCh8gCg+DaPB4ubXwGFvpuZcsJ\n\
         \UrpsV90T2M3ne5ie9+4BsowaC4KHBy8pBRgVMds3I0o8DqBGxCy6m/YTXIWhYDdV\n\
         \Lx24r0ilM1/48uSgPbZfd4RZHu1JSDMMfoqVZOcCgYEAnyu0oYpyUG4vCv64B+jS\n\
         \EeUu3ikUhCypcOmr8sXrmCOW8a0Mia6e3i9CFvExyxV4elxo7s9c0plf2oJ63HPb\n\
         \jmtJD9eqUw8bkqTXHdKxtMQnDIOIcSPjOm/LEEw8tBJDXlJu4PYKNMLuR8H5uzQQ\n\
         \UOJAMtEkd9feI1GJzcjDg/cCgYEAg4sLPfMeQi0kOpl0EEq0d4hvbhJJVU3n7Uyg\n\
         \ooW9ptoWRXIM9MIosSzkeBy6f2alc7oKEqX/SMK6Jr2OTNvyW/2r2+Jyh/IRKGeW\n\
         \b797e46Nx73nntFx3xo4KNczynaaK2z6S0AcUUcKTc0zznOuUdnOYxDzkGnPYKEs\n\
         \Uf5zAysCgYEAggHpi2g0SQi+GZjG9a4BbdDBiEX44V+BPf2LSd7bOvnC2YxGnppv\n\
         \hJcQR8jXO7NAmP/IVdk8X8smdfBGxjoQ8rmjUflYUEZCXbhPQLozc0nx7KWdjPKI\n\
         \sEHtYlvmmE0kZDbvco4caCATvX7pQgcpx6+chObTm6JLtYHoqcVjXPU=\n\
         \-----END RSA PRIVATE KEY-----"

serverCredential = right $ credentialLoadX509FromMemory certificatePem keyPem where
    right (Right cred) = cred
    right (Left err)   = error err

serverParams = def { serverShared = def { sharedCredentials = Credentials [serverCredential] }
                   , serverSupported = def { supportedCiphers = ciphersuite_default
                                           , supportedClientInitiatedRenegotiation = True }
                   }

startTlsServer :: TransportHandler -> IO TcpServer
startTlsServer handler = do
    svr <- newTlsServer serverParams listenPort handler
    waitListen svr
    return svr

noServerValidation = ValidationCache (\_ _ _ -> return ValidationCachePass) (\_ _ _ -> return ())

clientParams = ClientParams { clientUseMaxFragmentLength = Nothing
                            , clientServerIdentification = ("127.0.0.1", (BC8.pack . show) listenPort)
                            , clientUseServerNameIndication =False
                            , clientWantSessionResume = Nothing
                            , clientShared = def { sharedValidationCache = noServerValidation }
                            , clientHooks = def
                            , clientSupported = def { supportedCiphers = ciphersuite_default }
                            , clientDebug = def
                            }

connToTlsServer :: IO (Context, Socket)
connToTlsServer = do
    sk <- connToTcpServer
    ctx <- contextNew sk clientParams
    handshake ctx
    return (ctx, sk)

closeTls :: Context -> Socket -> IO ()
closeTls ctx sk = bye ctx `catch` (\(_ :: IOException) -> return ()) >> closeTcp sk

helloServerHandler :: ThreadMap -> Transport -> IO ()
helloServerHandler _ peer = transportSend peer helloWorldMessage

delayedHelloServerHandler :: ThreadMap -> Transport -> IO ()
delayedHelloServerHandler _ peer = do
    threadDelay (10 * 10^6)
    transportSend peer helloWorldMessage

echoServerHandler :: ThreadMap -> Transport -> IO ()
echoServerHandler _ peer = go
  where
    go = do
        msg <- transportRecv peer
        unless (null msg) $ do
            transportSend peer $ fromStrict msg
            go

spec :: Spec
spec = do
    describe "TCP based TcpServer with single shot return message" $ do
        it "accepts connection from client" $ do
            bracket (startTcpServer helloServerHandler) shutdownServer $ \_ -> do
                sk <- connToTcpServer
                isConnected sk >>= (`shouldBe` True)
                isWritable sk >>= (`shouldBe` True)
                close sk

        it "closes sending end after send a message" $ do
            bracket (startTcpServer helloServerHandler) shutdownServer $ \_ -> do
                sk <- connToTcpServer
                isConnected sk >>= (`shouldBe` True)
                isWritable sk >>= (`shouldBe` True)
                msg1 <- recv sk 4096
                fromStrict msg1 `shouldBe` helloWorldMessage
                msg2 <- recv sk 4096
                null msg2 `shouldBe` True
                close sk

        it "accepts multiple connection sequencially" $ do
            bracket (startTcpServer helloServerHandler) shutdownServer $ \_ -> do
                sk1 <- connToTcpServer
                isConnected sk1 >>= (`shouldBe` True)
                isWritable sk1 >>= (`shouldBe` True)
                close sk1
                sk2 <- connToTcpServer
                isConnected sk2 >>= (`shouldBe` True)
                isWritable sk2 >>= (`shouldBe` True)
                close sk2

        it "accepts multiple connection concurrently" $ do
            bracket (startTcpServer helloServerHandler) shutdownServer $ \_ -> do
                sk1 <- connToTcpServer
                sk2 <- connToTcpServer
                isConnected sk1 >>= (`shouldBe` True)
                isWritable sk1 >>= (`shouldBe` True)
                isConnected sk2 >>= (`shouldBe` True)
                isWritable sk2 >>= (`shouldBe` True)
                close sk1
                close sk2

    describe "TCP based TcpServer with delayed single shot return message" $
        it "forces disconnecting on server shutdown though handler has pending job" $ do
            svr <- startTcpServer delayedHelloServerHandler
            sk <- connToTcpServer
            threadDelay (10 * 10^3)
            shutdownServer svr
            threadDelay (10 * 10^3)
            msg1 <- recv sk 4096
            null msg1 `shouldBe` True
            close sk

    describe "TCP base EchoServer" $ do
        it "receives a message and echo back it" $ do
            bracket (startTcpServer echoServerHandler) shutdownServer $ \_ -> do
                sk <- connToTcpServer
                sendAll sk "hello"
                msg <- recv sk 4096
                msg `shouldBe` "hello"
                close sk

        it "echoes messages in arriving order" $ do
            bracket (startTcpServer echoServerHandler) shutdownServer $ \_ -> do
                sk <- connToTcpServer
                sendAll sk "hello, "
                sendAll sk "world"
                threadDelay (100 * 10^3)
                msg <- recv sk 4096
                msg `shouldBe` "hello, world"
                close sk

        it "receives and echoes messages in each session indipendently" $ do
            bracket (startTcpServer echoServerHandler) shutdownServer $ \_ -> do
                sk1 <- connToTcpServer
                sk2 <- connToTcpServer
                sendAll sk1 "hello"
                sendAll sk2 "world"
                msg1 <- recv sk1 4096
                msg2 <- recv sk2 4096
                msg1 `shouldBe` "hello"
                msg2 `shouldBe` "world"
                close sk1
                close sk2

        it "handles many sequencial sessions" $ do
            bracket (startTcpServer echoServerHandler) shutdownServer $ \_ -> do
                for_ [1..100] $ \n -> do
                    sk <- connToTcpServer
                    let smsg = BC8.pack $ show n
                    sendAll sk smsg
                    rmsg <- recv sk 4096
                    rmsg `shouldBe` smsg
                    close sk

        it "handles many concurrent sessions" $ do
            bracket (startTcpServer echoServerHandler) shutdownServer $ \_ -> do
                let smsgs = map (BC8.pack . show) [1..100]
                sks <- traverse (const connToTcpServer) [1..100]
                zipWithM_ sendAll sks smsgs
                rmsgs <- traverse (`recv` 4096) sks
                smsgs `shouldBe` rmsgs
                traverse_ close sks

    describe "TLS based TcpServer with single shot return message" $ do
        it "closes sending end after send a message" $ do
            bracket (startTlsServer helloServerHandler) shutdownServer $ \_ -> do
                (ctx, sk) <- connToTlsServer
                msg1 <- recvData ctx
                fromStrict msg1 `shouldBe` helloWorldMessage
                msg2 <- recvData ctx
                null msg2 `shouldBe` True
                closeTls ctx sk

        it "accepts multiple connection sequencially" $ do
            bracket (startTlsServer helloServerHandler) shutdownServer $ \_ -> do
                (ctx1, sk1) <- connToTlsServer
                contextGetInformation ctx1 >>= (`shouldSatisfy` isJust)
                closeTls ctx1 sk1
                (ctx2, sk2) <- connToTlsServer
                contextGetInformation ctx2 >>= (`shouldSatisfy` isJust)
                closeTls ctx2 sk2

        it "accepts multiple connection concurrently" $ do
            bracket (startTlsServer helloServerHandler) shutdownServer $ \_ -> do
                (ctx1, sk1) <- connToTlsServer
                (ctx2, sk2) <- connToTlsServer
                contextGetInformation ctx1 >>= (`shouldSatisfy` isJust)
                contextGetInformation ctx2 >>= (`shouldSatisfy` isJust)
                closeTls ctx1 sk1
                closeTls ctx2 sk2

    describe "TLS based TcpServer with delayed single shot return message" $
        it "forces disconnecting on server shutdown though handler has pending job" $ do
            svr <- startTlsServer delayedHelloServerHandler
            (ctx, sk) <- connToTlsServer
            threadDelay (10 * 10^3)
            shutdownServer svr
            threadDelay (10 * 10^3)
            msg1 <- recvData ctx
            null msg1 `shouldBe` True
            closeTls ctx sk

    describe "TLS base EchoServer" $ do
        it "receives a message and echo back it" $ do
            bracket (startTlsServer echoServerHandler) shutdownServer $ \_ -> do
                (ctx, sk) <- connToTlsServer
                sendData ctx "hello"
                msg <- recvData ctx
                msg `shouldBe` "hello"
                closeTls ctx sk

        it "echoes messages in arriving order" $ do
            bracket (startTlsServer echoServerHandler) shutdownServer $ \_ -> do
                (ctx, sk) <- connToTlsServer
                sendData ctx "hello, "
                sendData ctx "world"
                msg1 <- recvData ctx
                msg2 <- recvData ctx
                msg1 <> msg2 `shouldBe` "hello, world"
                closeTls ctx sk

        it "receives and echoes messages in each session indipendently" $ do
            bracket (startTlsServer echoServerHandler) shutdownServer $ \_ -> do
                (ctx1, sk1) <- connToTlsServer
                (ctx2, sk2) <- connToTlsServer
                sendData ctx1 "hello"
                sendData ctx2 "world"
                msg1 <- recvData ctx1
                msg2 <- recvData ctx2
                msg1 `shouldBe` "hello"
                msg2 `shouldBe` "world"
                closeTls ctx1 sk1
                closeTls ctx2 sk2

        it "handles many sequencial sessions" $ do
            bracket (startTlsServer echoServerHandler) shutdownServer $ \_ -> do
                for_ [1..100] $ \n -> do
                    (ctx, sk) <- connToTlsServer
                    let smsg = BLC8.pack $ show n
                    sendData ctx smsg
                    rmsg <- recvData ctx
                    fromStrict rmsg `shouldBe` smsg
                    closeTls ctx sk

        it "handles many concurrent sessions" $ do
            bracket (startTlsServer echoServerHandler) shutdownServer $ \_ -> do
                let smsgs = map (BLC8.pack . show) [1..100]
                ctxsSks <- traverse (const connToTlsServer) [1..100]
                let (ctxs, sks) = unzip ctxsSks
                zipWithM_ sendData ctxs smsgs
                rmsgs <- traverse recvData ctxs
                smsgs `shouldBe` map fromStrict rmsgs
                traverse_ (uncurry closeTls) ctxsSks


src/Network/TcpServer.hs
{-# LANGUAGE ScopedTypeVariables #-}


{-|
Module      : Network.TcpServer
Copyright   : (c) Naoto Shimazaki 2017
License     : MIT (see the file LICENSE)

Maintainer  : https://github.com/nshimaza
Stability   : experimental

A Simple Plain TCP Server.

Boot a simple, plain, general purpose TCP server with automatic hierarchical thread management.
It accepts a user provided handler and forks dedicated thread which listens given TCP port number.
Every time the listening thread accept a new connection from peer, it forks new thread dedicated for
each connection then calls user provided handler.

When server shutdown, all threads including listening thread, handler threads and their children
automatically killed.  You don't have to close the socket by your own.  This module automatically
close the socket on handler exit regardless if it exited normally or it was killed.
You can install your own cleanup routine using finally or bracket.


=== Example

@
import           Control.Monad        (forever, unless)
import qualified Data.ByteString      as B (null)
import qualified Data.ByteString.Lazy as BL (fromStrict)

import           Network.TcpServer

newEchoServer :: IO TcpServer
newEchoServer = newTlsServer 8443 echoServerHandler

echoServerHandler :: TransportHandler
echoServerHandler _ peer = go
  where
    go = do
        msg <- transportRecv peer
        unless (B.null msg) $ do
            transportSend peer $ BL.fromStrict msg
            go
@

-}
module Network.TcpServer
    (
    -- * Types
      TransportHandler
    , TcpServer
    , Transport
    -- * Functions
    , newTcpServer
    , newTlsServer
    , shutdownServer
    , waitListen
    , transportRecv
    , transportSend
    -- * Re-exported Types
    , ThreadMap
    ) where


import           Control.Concurrent.MVar        (MVar, newEmptyMVar, putMVar,
                                                 readMVar)
import           Control.Exception              (IOException, SomeException,
                                                 bracket, catch, finally,
                                                 throwIO)
import           Control.Monad                  (forever, void)
import           Data.ByteString                (ByteString)
import qualified Data.ByteString.Lazy           as L (ByteString, fromStrict)
import           Network.Socket                 (Family (..), PortNumber,
                                                 ShutdownCmd (..),
                                                 SockAddr (..), Socket,
                                                 SocketOption (..),
                                                 SocketType (..), accept, bind,
                                                 close, defaultProtocol,
                                                 iNADDR_ANY, listen,
                                                 setSocketOption, shutdown,
                                                 socket)
import qualified Network.Socket.ByteString      (recv)
import qualified Network.Socket.ByteString.Lazy (sendAll)
import           Network.TLS                    (Context, ServerParams (..),
                                                 bye, contextNew, handshake,
                                                 recvData, sendData)

import           Data.Default.Class

import           Control.Concurrent.Hierarchy   (ThreadMap, killThreadHierarchy,
                                                 newChild, newThreadMap)

{-|
    'Transport' is abstracted connection to remote peer.  The real backend can be bare TCP socket
    or TLS context.  'Transport' is passed to user provided connection handler.  User can read and write its
    contents via 'transportRecv' and 'transportSend' functions regardless which backend is actually used.
-}
data Transport = TransportSocket Socket | TransportTls Context

{-|
    Receive a chunk of data from the 'Transport'.  If the remote terminated the connection, zero length
    'ByteString' is returned.
-}
transportRecv :: Transport -> IO ByteString
transportRecv (TransportSocket sk) = Network.Socket.ByteString.recv sk 4096
transportRecv (TransportTls ctx)   = recvData ctx

{-|
    Send all content of the 'ByteString' to the 'Transport'.  The 'ByteString' must be lazy ByteString.
-}
transportSend :: Transport -> L.ByteString -> IO ()
transportSend (TransportSocket sk) msg  = Network.Socket.ByteString.Lazy.sendAll sk msg
transportSend (TransportTls ctx) msg    = sendData ctx msg

{-# DEPRECATED TcpHandler "use TransportHandler" #-}
{-|
    'TcpHandler' is type synonym of user provided function which actually working with single TCP connection.
    Your handler is called with 'ThreadMap' coming from thread-hierarchy package.
    You can ignore it as long as your handler doesn't create its own child thread.
    If the handler creates child, you must use 'newChild' in thread-hierarchy with passing the given 'ThreadMap'
    as the first argument of 'newChild' so that 'newChild' can properly install cleanup routine on exit of
    your handler or entire server.
-}
type TcpHandler
    =  ThreadMap    -- ^ Empty ThreadMap to be used when the handler creates its child thread with newChild.
    -> Socket       -- ^ TCP socket connected from peer which the handler working with.
    -> IO ()

{-|
    'TransportHandler' is type synonym of user provided function which actually working with single connection.
    The connection can be nicked TCP connection or already established TLS connection.
    Your handler is called with 'ThreadMap' coming from thread-hierarchy package.
    You can ignore it as long as your handler doesn't create its own child thread.
    If the handler creates child, you must use 'newChild' in thread-hierarchy with passing the given 'ThreadMap'
    as the first argument of 'newChild' so that 'newChild' can properly install cleanup routine on exit of
    your handler or entire server.
-}
type TransportHandler
    =  ThreadMap    -- ^ Empty ThreadMap to be used when the handler creates its child thread with newChild.
    -> Transport    -- ^ Abstracted connection to peer which the handler working with.
    -> IO ()

makeTcpHandler :: TransportHandler -> (ThreadMap -> Socket -> IO ())
makeTcpHandler handler handlerChildren peer = handler handlerChildren $ TransportSocket peer

makeTlsHandler :: ServerParams -> TransportHandler -> (ThreadMap -> Socket -> IO ())
makeTlsHandler params handler handlerChildren peer =
    bracket
        (do
            ctx <- contextNew peer params
            handshake ctx
            return ctx)
        (\ctx ->
            -- If an async exception was thrown, say killThread, we still have active TLS context.
            -- So we need to try to gracefully shutdown it.
            -- However, there is some possibility that the context was already invalidated when we reached here.
            -- For example, if user provided handler returned because it detected EoF on receiving data,
            -- the context can be already invalid and close notification may result error.
            -- Causing IOException by trying to close already invalid context is not a problem for us,
            -- so we just catch the exception then ignore it.
            bye ctx `catch` \(_ :: IOException) -> return ()
        )
        (handler handlerChildren . TransportTls)

{-|
    A simple plain TCP server.
-}
data TcpServer = TcpServer ThreadMap (MVar ())

{-|
    Create a new 'TcpServer' with non-secure transport.  It forks a new thread to listen given TCP port.
    It calls user provided 'TransportHandler' every time it accepts new TCP connection from peer.
    The 'TransportHandler' is executed by its dedicated thread.
-}
newTcpServer
    :: PortNumber       -- ^ TCP port number the newly created server to listen.
    -> TransportHandler -- ^ User provided function handling abstracted transport.
    -> IO TcpServer     -- ^ newTcpSever returns created server object.
newTcpServer port handler = newServer port $ makeTcpHandler handler

{-|
    Create a new 'TcpServer' with TLS transport.  It forks a new thread to listen given TCP port.
    It calls user provided 'TransportHandler' every time it accepts new TCP connection from peer.
    The 'TransportHandler' is executed by its dedicated thread.
-}
newTlsServer
    :: ServerParams     -- ^ Parameters for TLS server.
    -> PortNumber       -- ^ TCP port number the newly created server to listen.
    -> TransportHandler -- ^ User provided function handling abstracted transport.
    -> IO TcpServer     -- ^ newTlsSever returns created server object.
newTlsServer params port handler = newServer port $ makeTlsHandler params handler


{-# DEPRECATED newServer "use newTcpServer or newTlsServer" #-}
{-|
    Create a new 'TcpServer'.  It forks a new thread to listen given TCP port.
    It calls user provided 'TcpHandler' every time it accepts new TCP connection from peer.
    The 'TcpHandler' is executed by its dedicated thread.
-}
newServer
    :: PortNumber   -- ^ TCP port number the newly created server to listen.
    -> TcpHandler   -- ^ User provided function handling each TCP connection.
    -> IO TcpServer -- ^ newSever returns created server object.
newServer port handler = do
    readyToConnect <- newEmptyMVar
    rootThreadMap <- newThreadMap
    void . newChild rootThreadMap $ \listenerChildren ->
        bracket (newListener port readyToConnect) (close) (acceptLoop listenerChildren handler)
    return $ TcpServer rootThreadMap readyToConnect

{-|
    Wait for given 'TcpServer' ready to accept new connection from peer.
    This is useful for unit testing where you can synchronize client connection to server boot up.
-}
waitListen :: TcpServer -> IO ()
waitListen (TcpServer _ readyToConnect) = readMVar readyToConnect >>= \_ -> return ()

{-|
    Shutdown given 'TcpServer'.  It kills all threads of its children, grandchildren and so on.
-}
shutdownServer :: TcpServer -> IO ()
shutdownServer (TcpServer rootThreadMap _) = killThreadHierarchy rootThreadMap


newListener :: PortNumber -> MVar() -> IO Socket
newListener port readyToConnect = do
    sk <- socket AF_INET Stream defaultProtocol
    setSocketOption sk ReuseAddr 1
    bind sk (SockAddrInet port iNADDR_ANY)
    listen sk 10000
    return sk
    putMVar readyToConnect ()
    return sk

acceptLoop :: ThreadMap -> TcpHandler -> Socket -> IO ()
acceptLoop listenerChildren handler sk = forever $ do
    (peer, _) <- accept sk
    void . newChild listenerChildren $ \handlerChildren ->
        handler handlerChildren peer `finally` (do
            -- When we reached here, there are two possible case.
            -- (1) We received asynchronous exception
            -- In this case, we should try to shutdown the socket gracefully.
            -- (2) User provided handler returned because it has no remaining tasks (e.g. EoF detected)
            -- In this case, the socket can be terminated by remote.  Trying to shutdown the socket may
            -- fail but we can ignore such failure.
            shutdown peer ShutdownSend `catch` (\(_ :: IOException) -> return ())
            close peer)
